---
title: "V1 APP + Base Genome Benchmark Analysis"
author: "Jaclyn Beck (Sage Bionetworks)"
output:
  html_document: 
    code_folding: hide
    highlight: pygments
    embed-resources: true
    toc: true
    df_print: default
    toc_depth: 1
  md_document: 
    pandoc_args: ["--wrap=preserve"]
---

```{r optional-wiki-setup, echo=FALSE, results="asis", eval=FALSE}
# Change this code chunk to "eval=TRUE" when knitting to a Synapse wiki to
# remove code from the render. This also makes the plots print out in higher
# quality than default for the wiki.

options(knitr.table.format = "pipe")
knitr::opts_chunk$set(echo = FALSE)

# This resolution and aspect ratio, scaled to ~45% on the wiki, looks good
knitr::opts_chunk$set(dpi = 300)
knitr::opts_chunk$set(fig.height = 4)
knitr::opts_chunk$set(fig.width = 7)

# Prints out a markdown header for the Synapse wiki and a reminder to download
# the HTML file for inline code display
cat("#! V1 APP + Base Genome Benchmark Analysis\n\n")
cat("**Author:** Jaclyn Beck (Sage Bionetworks)\n\n")
cat("> **Note:** This wiki page shows the output of the R notebook but with",
    "inline code removed. **Download the rendered HTML file** with the",
    "`Download Options` button to view the notebook with code inserted.",
    "\n\n", "${toc}", "\n\n") # wiki table of contents
```

# Methods

I generated a custom genome using the methods described in [V1 Genome Benchmark Analysis](https://www.synapse.org/Synapse:syn73455492), however only human APP was added to the base mouse genome. All samples were aligned to the APP+ genome, and compared to the alignment with the base genome.

This benchmarking case tries to isolate the effect of adding the human APP gene on alignment.

# Setup

Read in metadata, processed counts files, and gene symbol file. Subset metadata to only the samples present in the counts files.

```{r message=FALSE, warning=FALSE, results='hide'}

library(stringr)
library(dplyr)
library(reshape2)
library(ggplot2)
library(RColorBrewer)
library(synapser)
library(patchwork)
library(knitr)
library(kableExtra)

setwd("..")

synLogin(silent = TRUE)

# Use V3 of the metadata, which uses the old Jax specimen IDs corresponding to
# what is in the benchmark files
meta_file <- synGet("syn61850266.3", downloadLocation = file.path("data", "tmp"))
metadata <- read.csv(meta_file$path)

# Necessary because R puts an "X" in front of any specimen IDs that start with
# a number to make column names
metadata$specimenID_X <- make.names(metadata$specimenID)

# Use V1 of the benchmark files, which do not include CLU tests
base_file <- synGet("syn61806059.1", downloadLocation = file.path("data", "tmp"))
counts_base <- read.table(base_file$path, sep = "\t", header = TRUE) |>
  select(-transcript_id.s.)

app_file <- synGet("syn61806423.1", downloadLocation = file.path("data", "tmp"))
counts_app <- read.table(app_file$path, sep = "\t", header = TRUE) |>
  select(-transcript_id.s.)

metadata <- subset(metadata, specimenID_X %in% colnames(counts_base)) |>
  select(individualID, specimenID, specimenID_X, genotype, study_name) |>
  # Shorten "APOE4-KI_homozygous; Trem2-R47H_homozygous" genotype to "LOAD1"
  mutate(genotype = ifelse(genotype == "APOE4-KI_homozygous; Trem2-R47H_homozygous",
                           "LOAD1", genotype))

symbol_file <- synGet("syn62063692", downloadLocation = file.path("data", "tmp"))
gene_symbols <- read.csv(symbol_file$path)
```

Set up plot themes

```{r}
bold_labels <- theme(title = element_text(face = "bold"),
                     axis.title = element_text(face = "bold"))

legend_bottom <- theme(legend.position = "bottom")

angled_x_axis <- theme(axis.text.x = element_text(angle = 90, hjust = 1, 
                                                  vjust = 0.5))

left_margin <- theme(plot.margin = unit(c(0, 0, 0, 50), units = "pt"))

facet_style <- theme(strip.background = element_rect(fill = "#EFEFEF"),
                     strip.text = element_text(face = "bold"))
```

# Quality checks

Check that:

-   The APP+ genome contains human genes, and the base genome doesn't
-   All human genes start with "ENSG" and all mouse genes start with "ENSM"
-   No genes are in the base genome that don't exist in the APP+ genome
-   All samples are present in both matrices

```{r}
# Check for existence of human genes
human_genes <- setdiff(counts_app$gene_id, counts_base$gene_id)
mouse_genes <- intersect(counts_base$gene_id, counts_app$gene_id)

cat("Genes present in APP+ genome that are not in base genome:", "\n",
    paste(human_genes, collapse = ", "), "\n")

# Naming of human and mouse genes
stopifnot(all(str_detect(human_genes, "ENSG")))
cat("All human genes start with 'ENSG'.", "\n")
stopifnot(all(str_detect(mouse_genes, "ENSM")))
cat("All mouse genes start with 'ENSM'", "\n")

# All mouse genes present in both genomes
stopifnot(length(setdiff(counts_base$gene_id, counts_app$gene_id)) == 0)
cat("All genes in base genome are present in APP+ genome.", "\n")

# All samples present
stopifnot(all(colnames(counts_base) %in% colnames(counts_app)))
cat("All samples are present in both counts matrices.", "\n")
```

# How many counts are identical or different between matrices?

Ensure that both matrices have samples in the same order, and create data frames that only include mouse genes for this analysis.

```{r message=FALSE}
rownames(counts_base) <- counts_base$gene_id
rownames(counts_app) <- counts_app$gene_id
counts_app <- counts_app[, colnames(counts_base)]

df_base <- melt(counts_base[mouse_genes, ], 
                variable.name = "sample", 
                value.name = "count")
df_app <- melt(counts_app[mouse_genes, ], 
                variable.name = "sample", 
                value.name = "count")
```

Some stats on identical counts vs mismatched counts:

```{r}
pct <- 100 * sum(df_base$count == df_app$count) / nrow(df_base)
cat("Identical counts between matrices:", 
    sum(df_base$count == df_app$count), 
    "of", nrow(df_base), 
    paste0("(", round(pct, digits = 2), "%)"), "\n")

cat("Mismatched counts between matrices:", 
    sum(df_base$count != df_app$count), 
    "of", nrow(df_base), 
    paste0("(", round(100-pct, digits = 2), "%)"), "\n")

df_combined <- subset(df_base, df_base$count != df_app$count) |>
  merge(df_app, by = c("gene_id", "sample")) |>
  dplyr::rename(count_base = "count.x", count_app = "count.y") |>
  mutate(difference = count_base - count_app)

cat("Correlation between mismatches: R^2 =", 
    cor(df_combined$count_base, df_combined$count_app)^2, "\n")

cat("Mean mismatch difference:", 
    mean(abs(df_combined$difference)), "counts \n")
```

## How many entries have a difference of \> 1 count?

RSEM returns decimal numbers when reads are ambiguous, so the differences are sometimes not integers. We consider a count difference \<= 1 to be trivial and ignore it in further analysis.

```{r}
df_combined <- merge(df_combined, metadata, by.x = "sample", by.y = "specimenID_X")

diff_counts_gt1 <- which(abs(df_combined$difference) > 1)

cat(length(diff_counts_gt1), "of", nrow(df_combined), 
    "differences are greater than 1 count.")
```

# Examining mouse gene count differences for differences \> 1 count

## By study

```{r}
ggplot(df_combined[diff_counts_gt1, ], 
       aes(x = count_base+1, y = abs(difference), color = study_name)) + 
  geom_point(size = 0.5) + scale_x_log10() + scale_y_log10() + 
  theme_bw() + legend_bottom + bold_labels +
  scale_color_brewer(palette = "Set1", name = "Study") +
  ggtitle("Count differences between reference genomes (for differences > 1)") + 
  xlab("Counts (aligned with base mouse genome)") + 
  ylab("Difference from APP+ genome")
```

## By genotype

Note that this graph appears to be dominated by the *5XFAD_carrier* genotype, and difference increases with total count.

```{r}
ggplot(df_combined[diff_counts_gt1, ], 
       aes(x = count_base+1, y = abs(difference), color = genotype)) + 
  geom_point(size = 0.5) + scale_x_log10() + scale_y_log10() + 
  theme_bw() + 
  legend_bottom + bold_labels +
  theme(legend.text = element_text(size = 6)) + 
  guides(color = guide_legend(nrow = 3, keywidth = 0.2)) + 
  scale_color_brewer(palette = "Dark2", name = "Genotype") + 
  ggtitle("Count differences between reference genomes (for differences > 1)") + 
  xlab("Counts (aligned with base mouse genome)") + 
  ylab("Difference from APP+ genome") 
```

## Histograms: Genes vs Samples

For each sample, how many genes have a count difference \> 1 between the base mouse genome and the APP+ genome?

```{r}
diff_samps <- table(df_combined$specimenID[diff_counts_gt1])
diff_samps <- as.data.frame(diff_samps) |>
  rename(Sample = Var1, Genes = Freq)

plt_g_per_s <- ggplot(diff_samps, aes(x = Genes)) + 
  geom_histogram(bins = 20, color = "black", fill = "lightskyblue") + 
  theme_bw() + xlab("Number of different genes") +
  bold_labels +
  ggtitle("Number of different genes \nper sample")
```

For each gene, how many samples have a count difference \> 1 between the base mouse genome and the APP+ genome?

```{r}
diff_genes <- table(df_combined$gene_id[diff_counts_gt1])
diff_genes <- as.data.frame(diff_genes) |>
  rename(Gene = Var1, Samples = Freq)

plt_s_per_g <- ggplot(diff_genes, aes(x = Samples)) + 
  geom_histogram(bins = 20, color = "black", fill = "thistle") + 
  theme_bw() + xlab("Number of different samples") +
  ggtitle("Number of different samples \nper gene") + 
  bold_labels + left_margin
```

```{r}
print(plt_g_per_s + plt_s_per_g)
```

## Histogram: differences by genotype

For each genotype, how many entries in the matrix (over all samples and genes) have a count difference \> 1 between the base mouse genome and the APP+ genome?

```{r}
diff_geno <- table(df_combined$genotype[diff_counts_gt1])
diff_geno <- as.data.frame(diff_geno) |>
  rename(Genotype = Var1, Entries = Freq)

plt_e_per_g <- ggplot(diff_geno, aes(x = Genotype, y = Entries)) + 
  geom_col(color = "black", fill = "darkseagreen3") + 
  theme_bw() + ylab("Matrix entries") +
  bold_labels + angled_x_axis +
  ggtitle("Number of different matrix \nentries per genotype")
```

For each genotype, how many unique genes have a count difference \> 1 in any sample between the base mouse genome and the APP+ genome?

```{r}
diff_geno2 <- df_combined[diff_counts_gt1, ] |> 
  select(gene_id, genotype) |>
  distinct()
diff_geno2 <- table(diff_geno2$genotype)
diff_geno2 <- as.data.frame(diff_geno2) |>
  rename(Genotype = Var1, Genes = Freq)

plt_g_per_g <- ggplot(diff_geno2, aes(x = Genotype, y = Genes)) + 
  geom_col(color = "black", fill = "salmon2") + 
  theme_bw() +
  bold_labels + angled_x_axis + left_margin +
  ggtitle("Number of unique genes \nchanged per genotype")
```

```{r}
print(plt_e_per_g + plt_g_per_g)
```

# Focusing on differences of \> 10 counts

Note that from the graphs above, the only entries with \> 10 counts of difference between the base mouse genome and the APP+ genome are entries with between 1,000 and 1,000,000 total counts, representing a margin of 1% to 0.001% difference.

## Which genotypes have more than a 10 count difference for any gene or sample?

```{r}
df_tmp <- subset(df_combined, abs(difference) > 10) |>
  group_by(genotype) |>
  summarize(n_differences = n())

kable(df_tmp) |> kable_styling(full_width = FALSE)
```

## How many unique genes is this?

```{r}
df_tmp <- subset(df_combined, abs(difference) > 10) |>
  group_by(gene_id) |>
  summarize(n_differences = n())

cat(nrow(df_tmp), "unique genes have at least one sample with a",
    "> 10 count difference.", "\n")
```

# Examining humanized gene expression

## Setup

Create a data frame with gene expression from humanized genes and mouse orthologs from the counts aligned with the APP+ genome, and the mouse orthologs from the counts aligned with the base genome. Give human genes and their mouse orthologs recognizable gene names.

```{r message=FALSE}
id_map <- c("ENSMUSG00000022892" = "APP",
            "ENSG00000142192" = "APP")

base_tmp <- melt(counts_base[names(id_map)[1], ],
                 variable.name = "sample", 
                 value.name = "count") |>
  mutate(gene_source = "Mouse App (base genome)")

df_human <- melt(counts_app[names(id_map), ], 
                 variable.name = "sample", 
                 value.name = "count") |>
  mutate(gene_source = "Mouse App (APP+ genome)") |>
  rbind(base_tmp) |>
  merge(metadata, by.x = "sample", by.y = "specimenID_X")

df_human$gene_name <- id_map[df_human$gene_id]
df_human$gene_source[grepl("ENSG", df_human$gene_id)] <- "Human APP"
```

Mark which genotypes have which humanized genes, for plotting.

```{r}
df_human$humanized_APP <- df_human$genotype %in% c("3xTg-AD_carrier",
                                                   "5XFAD_carrier")
```

## Gene expression

*3xTg-AD_carrier* and *5XFAD_carrier* mice express humanized APP. hAPP was added to the genome with no alterations to the mouse *App* gene, so it is expected that levels of mouse *App* for these mice would be comparable to noncarrier genotypes.

Some mice from the *3xTg-AD_noncarrier*, *APOE4-KI_noncarrier*, and *5XFAD_noncarrier* genotypes, which do **not** contain hAPP, had a small number of detected counts (\<= 3 counts) for humanized APP. One *5XFAD_noncarrier* sample had 533 detected counts, and this sample was mis-genotyped or contaminated so we do not consider the detection an issue for this genotype.

```{r}
app_counts <- subset(df_human, gene_id == "ENSG00000142192") |>
  group_by(genotype) |>
  summarize(max_hAPP = round(max(count)),
            n_detected_hAPP = sum(count > 0))
kable(app_counts) |> kable_styling(full_width = FALSE)
```

```{r, fig.height=6}
ggplot(df_human, aes(x = genotype, y = count+1, color = humanized_APP)) +
  geom_jitter(width = 0.2) + scale_y_log10(na.value = 0) + 
  facet_wrap(~gene_source) +
  theme_bw() + 
  ylab("count") + 
  scale_color_manual(values = c("darkgray", "deepskyblue3"),
                     name = "Has humanized APP") +
  bold_labels + angled_x_axis + legend_bottom + facet_style +
  ggtitle("Gene expression, emphasis on genotypes with hAPP")
```

# Which mouse genes specifically are changed between references?

```{r warning=FALSE}
gene_symbols <- subset(gene_symbols, 
                       ensembl_gene_id %in% df_combined[diff_counts_gt1, "gene_id"])

df_per_geno <- df_combined[diff_counts_gt1,] |> 
  group_by(genotype) |>
  summarize(gene_ids = list(unique(gene_id)))

for (geno in df_per_geno$genotype) {
  ids <- df_per_geno$gene_ids[df_per_geno$genotype == geno][[1]]
  genes <- subset(gene_symbols, ensembl_gene_id %in% ids)
  cat(geno, "\n")
  print(sort(genes$gene_symbol))
  cat("\n\n")
}
```

Filter out non-coding RNA, mitochondrial genes, ribosomal genes, un-annotated genes, and pseudogenes from this list.

```{r}
for (geno in df_per_geno$genotype) {
  ids <- df_per_geno$gene_ids[df_per_geno$genotype == geno][[1]]
  genes <- subset(gene_symbols, ensembl_gene_id %in% ids & gene_symbol != "")
  genes <- genes[!grepl("-ps|mt-|^Rp|^Rn[0-9]|^Gm[0-9]+|^[0-9]+", genes$gene_symbol), ]
  cat(geno, "\n")
  print(sort(genes$gene_symbol))
  cat("\n\n")
}
```

**What are the 18 genes with count differences \> 10?**

```{r}
df_tmp <- subset(df_combined, abs(difference) > 10) |>
  group_by(gene_id) |>
  summarize(n_differences = n())

genes <- subset(gene_symbols, ensembl_gene_id %in% df_tmp$gene_id)
print(sort(genes$gene_symbol))
```

Expression differences of these genes appear to be near-direct swaps between two genes within the same sample:

```{r }
df_tmp2 <- subset(df_combined, abs(difference) > 10) |>
  merge(gene_symbols, by.x = "gene_id", by.y = "ensembl_gene_id") |>
  select(gene_id, gene_symbol, difference, specimenID, genotype) |>
  subset(gene_symbol != "App") |>
  arrange(specimenID, abs(difference))

kable(df_tmp2) |> kable_styling(full_width = FALSE) |> scroll_box(height = "400px")
```

# Conclusion

-   Detection of human APP appears consistent with each mouse genotype, except for the one potentially contaminated *5XFAD_noncarrier* sample.
-   Detection of mouse *App* is similar between the base mouse genome and the APP+ genome, and expression levels are as expected across genotypes.
-   However, adding humanized genes to the reference *does change detected counts for other mouse genes* as well, excluding mouse *App*:
    -   the majority of these genes are pseudogenes, mitochondrial, ribosomal, un-annotated, or non-coding RNAs
    -   Looking at all of the benchmarking for human genes individually, it's clear that **adding APP contributes to most of the variation** found between the full custom genome and the base genome.
    -   **It's unclear to me how much this matters**, although the large number of affected genes in 5XFAD carriers is potentially concerning.
        -   However it looks like most of the large changes are swapped counts between a gene and its pseudogene, or between a gene and a related unannotated gene, or between non-coding RNAs. **This seems like it would make an insignificant impact on analysis.**

```{r}
# Get some statistics on the scope of gene changes per genotype
df_gt1 <- df_combined[diff_counts_gt1,] |>
  subset(!(gene_id %in% names(id_map))) # Remove human genes + mouse orthologs
df_gt1$control <- grepl("noncarrier", df_gt1$genotype)
df_gt1$control[df_gt1$genotype == "Trem2-R47H_homozygous"] = TRUE
df_gt1$apoe <- grepl("APOE4-KI_homozygous", df_gt1$genotype)

stats_by_genotype <- df_gt1 |> group_by(control, genotype) |> 
  summarize(n_genes = length(unique(gene_id)), 
            median = median(abs(difference)), 
            max = max(abs(difference)),
            .groups = "drop")
kable(stats_by_genotype) |> kable_styling(full_width = FALSE)
```

```{r upload, eval=FALSE, include=FALSE}
# Code to add this notebook's rendered contents to a wiki page on Synapse, which
# makes it nicely viewable. The wiki belongs to the rendered HTML file that was
# uploaded to Synapse.
#
# NOTE: Images still need to be re-scaled on Synapse at this time because the
# knit2synapse library currently assumes the scaling should always be 100%. The
# yaml header also needs to be manually removed from the page.
library(knit2synapse)
library(synapser)

synLogin(silent = TRUE)

knitfile2synapse(file = "V1c_APP_Benchmark_Analysis.Rmd",
                 owner = "syn73461438",
                 wikiName = "V1c APP + Base Genome Benchmark Analysis")
```
