---
title: "Jax.IU.Pitt_5XFAD Dataset Exploration"
author: "Jaclyn Beck (Sage Bionetworks)"
date: "`r Sys.Date()`"
output:
  html_document: 
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r initialize, include=FALSE}
source("vis_utils.R")
data <- load_data("Jax.IU.Pitt_5XFAD")

# "data" includes the following items:
#   metadata, counts = data frames containing only valid samples
#   all_metadata, all_counts = data frames containing valid + invalid samples
#   all_counts_norm = the all_counts matrix normalized with vst() for PCA plotting

# For nicer printing, add "months" to the end of ageDeath
meta_vis <- data$metadata %>%
  mutate(ageDeath = paste(ageDeath, ageDeathUnits))
```

## Sample validation

```{r print-invalid-samples}
print_invalid_samples(data$all_metadata,
                      columns_print = c("valid_5x_variant", "valid_5x_expression"))
```

PCA of all samples to see where the mismatch clusters.

```{r mismatch-pca, message=FALSE}
create_pca_plot(data$all_metadata, data$all_counts_norm,
                title = "PCA of all samples",
                # aes specification
                color = valid_sample, shape = genotype)
```

Expression of APP and PSEN1 in all carrier samples plus the mismatched sample to see how the mismatched non-carrier compares.

```{r carrier-expression}
carriers <- subset(data$all_metadata,
                   genotype == "5XFAD_carrier" | !valid_sample)
carriers <- carriers$merged_file_specimenID

plot_gene_counts(data$all_metadata, data$all_counts[, carriers],
                 genes_plot = c("APP", "PSEN1"),
                 draw_boxplot = FALSE,
                 title = "Expression in carriers vs mismatched non-carrier",
                 # aes specification
                 x = valid_sample, y = counts, color = valid_sample)
```

Expression of APP and PSEN1 in all non-carrier samples to see how the mismatched non-carrier compares.

```{r nc-expression}
noncarriers <- subset(data$all_metadata, genotype == "5XFAD_noncarrier")
noncarriers <- noncarriers$merged_file_specimenID

plot_gene_counts(data$all_metadata, data$all_counts[, noncarriers],
                 genes_plot = c("APP", "PSEN1"),
                 draw_boxplot = FALSE,
                 title = "Expression in non-carriers",
                 # aes specification
                 x = valid_sample, y = counts, color = valid_sample)
```

This mismatched sample is borderline: all 6 5X variants were detected in this sample via genotype calling, but it expresses APP and PSEN1 at far lower levels than any other carrier sample. However, the expression of both genes, APP especially, is noticeably higher in this sample than all other non-carriers in the data set too. Given the ambiguity, we exclude this sample to be safe. 

## Study design

The 11-month cohort are technically supposed to be 12 months but all mice are marked with an ageDeath close to 11. 

```{r study-design}
print_study_design(meta_vis, "sex", "genotype", "ageDeath")
```

## Check for batches

No batches present:

```{r batches}
data$metadata %>%
  group_by(rnaBatch, libraryBatch, sequencingBatch) %>%
  summarize(count = n(), .groups = "drop")
```

## Human gene expression

```{r human-expression, fig.width=8}
plot_gene_counts(data$metadata, data$counts,
                 genes_plot = c("APOE", "APP", "MAPT", "PSEN1"),
                 title = "Expression of human transgenes",
                 # aes specification
                 x = genotype, y = counts, color = genotype)
```

## Quality plots of samples

```{r deseq, message=FALSE, warning=FALSE}
keep <- rowSums(data$counts > 10) >= 3
dds <- DESeqDataSetFromMatrix(data$counts[keep, ], meta_vis, design = ~ 1)
dds <- DESeq(dds, quiet = TRUE)
plotDispEsts(dds)
```

```{r vst}
vsd <- vst(dds)
meanSdPlot(assay(vsd))
```

## PCA plots

```{r pca-plots, message=FALSE, warning=FALSE}
plotPCA(vsd, intgroup = c("genotype", "sex")) +
  ggtitle("Genotype vs sex") +
  scale_color_brewer(palette = "Set1") + theme_bw()

plotPCA(vsd, intgroup = c("genotype", "ageDeath")) +
  ggtitle("Genotype vs age") +
  scale_color_brewer(palette = "Set1") + theme_bw()

plotPCA(vsd, intgroup = c("sex", "ageDeath")) +
  ggtitle("Sex vs age") +
  scale_color_brewer(palette = "Set1") + theme_bw()
```

5XFAD samples cluster by genotype, sex, and age. Age differences are most pronounced in 5XFAD carriers.

```{r upload, eval=FALSE, include=FALSE}
# Code to upload this file to Synapse and add its contents to its own wiki page,
# which makes it nicely viewable.
library(knit2synapse)
library(synapser)

synLogin(silent = TRUE)

# TODO add provenance
createAndKnitToFileEntity(file = "Jax.IU.Pitt_5XFAD_dataset.Rmd",
                          parentId = "syn64299959",
                          fileName = "Jax.IU.Pitt_5XFAD_dataset.Rmd",
                          wikiName = "Jax.IU.Pitt_5XFAD Dataset Exploration",
                          overwrite = TRUE,
                          forceVersion = FALSE)
```
