---
title: "UCI_5XFAD Dataset Exploration"
author: Jaclyn Beck (Sage Bionetworks)
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Set up

```{r initialize, include=FALSE}
source("vis_utils.R")
data <- load_data("UCI_5XFAD")

# "data" includes the following items:
#   metadata, counts = data frames containing only valid samples
#   all_metadata, all_counts = data frames containing valid + invalid samples
#   all_counts_norm = the all_counts matrix normalized with vst() for PCA plotting

# For nicer printing, add "months" to the end of ageDeath
meta_vis <- data$metadata %>%
  mutate(ageDeath = paste(ageDeath, ageDeathUnits))
```

## Sample validation

```{r print-invalid-samples}
print_invalid_samples(data$all_metadata,
                      columns_print = c("valid_5x_variant", "valid_5x_expression"))
```

PCA of all samples to see where the genotype mismatches cluster.

```{r mismatch-pca, message=FALSE}
data$all_metadata <- data$all_metadata %>%
  mutate(valid_genotype = ifelse(valid_sex & !valid_sample, FALSE, TRUE))

create_pca_plot(data$all_metadata, data$all_counts_norm,
                title = "PCA of all samples",
                # aes specification
                color = valid_genotype, shape = genotype)
```

Sex-related gene expression

```{r sex-counts-plot}
data$all_metadata <- data$all_metadata %>%
  mutate(sex = as.character(sex),
         sex = ifelse(valid_sex, sex, paste(sex, "mismatch")))

plot_gene_counts(data$all_metadata, data$all_counts,
                 genes_plot = c("Xist", "Eif2s3y", "Ddx3y"),
                 draw_boxplot = FALSE,
                 title = "Sex-related gene expression",
                 # aes specification
                 x = sex, y = counts, color = sex, shape = sex)
```

All 8 samples (from 4 mice) are clear mismatches and will be excluded.

## Study design

```{r study-design}
print_study_design(meta_vis, c("sex", "tissue"), "genotype", "ageDeath")
```

## Check for batches

There are 4 batches present in the data:

```{r batches}
data$metadata %>%
  group_by(rnaBatch, libraryBatch, sequencingBatch) %>%
  summarize(count = n(), .groups = "drop")
```

It is unclear whether the `NA` batches represent missing information or a separate batch. For analysis purposes we will treat it as a separate batch.

All 3 batch variables are identical.

Batches contain both genotypes:

```{r batches2}
meta_vis <- meta_vis %>%
  mutate(batch = paste("Batch", rnaBatch))

table(meta_vis$batch, meta_vis$genotype)
```

Batches contain both sexes:

```{r batches3}
table(meta_vis$batch, meta_vis$sex)
```

Batches are highly specific to `ageDeath`:

```{r batches4}
table(meta_vis$batch, meta_vis$ageDeath)
```

The single 12-month sample in the `NA` batch could maybe be thrown out.

Batches contain both tissues:

```{r batches5}
table(meta_vis$batch, meta_vis$tissue)
```

Breakdown of batches within each analysis group

```{r batches6}
print_study_design(meta_vis, c("sex", "tissue", "ageDeath"), "genotype", "batch")
```

For male 18 month mice (both tissues), there is only 1 sample per tissue/genotype in the `NA` batch. The male 12 month mice (cerebral cortex) have the same issue. This could suggest that at least some of the `NA` batch labels are missing data instead of a separate batch. 

The 18-month sample sizes for both sexes are very unbalanced between groups, so it's unclear whether this group should even be analyzed, and if so, whether batch differences will negatively effect the male samples.

## Human gene expression

```{r human-expression, fig.width=8}
plot_gene_counts(data$metadata, data$counts,
                 genes_plot = c("APOE", "APP", "MAPT", "PSEN1"),
                 title = "Expression of human transgenes",
                 # aes specification
                 x = genotype, y = counts, color = genotype)
```

## Quality plots of samples -- Cerebral cortex

```{r dds-cc, message=FALSE, warning=FALSE}
meta_cc <- subset(meta_vis, tissue == "cerebral cortex")
counts_cc <- data$counts[, meta_cc$merged_file_specimenID]

keep <- rowSums(counts_cc > 10) >= 3

dds_cc <- DESeqDataSetFromMatrix(counts_cc[keep, ], meta_cc, design = ~ 1)
dds_cc <- DESeq(dds_cc, quiet = TRUE)
plotDispEsts(dds_cc)
```

```{r vst-cc}
vsd_cc <- vst(dds_cc)
meanSdPlot(assay(vsd_cc))
```

## PCA plots -- Cerebral cortex

```{r pca-cc, message=FALSE, warning=FALSE}
# DESeq2's plotPCA function forces numeric variables to be categorical, so we
# use my function for plotting instead
meta_cc$RIN <- as.numeric(meta_cc$RIN)
create_pca_plot(meta_cc, assay(vsd_cc),
                title = "RIN",
                discrete_color = FALSE,
                brewer_palette = "OrRd",
                # aes specification
                color = RIN)

# Boxplots of RIN vs each group
meta_cc$group <- paste(meta_cc$sex, meta_cc$ageDeath)
plot_grouped_boxplot(meta_cc,
                     title = "RIN vs group",
                     # aes specification
                     x = group, y = RIN, color = genotype)

plotPCA(vsd_cc, intgroup = c("genotype", "sex")) +
  ggtitle("Genotype vs sex") +
  scale_color_brewer(palette = "Set1") + theme_bw()

plotPCA(vsd_cc, intgroup = c("genotype", "ageDeath")) +
  ggtitle("Genotype vs age") +
  scale_color_brewer(palette = "Set1") + theme_bw()

plotPCA(vsd_cc, intgroup = c("sex", "ageDeath")) +
  ggtitle("Sex vs age") +
  scale_color_brewer(palette = "Set1") + theme_bw()

plotPCA(vsd_cc, intgroup = c("rnaBatch", "ageDeath")) +
  ggtitle("Batch vs age") +
  scale_color_brewer(palette = "Set1") + theme_bw()

plotPCA(vsd_cc, intgroup = c("rnaBatch", "genotype")) +
  ggtitle("Batch vs genotype") +
  scale_color_brewer(palette = "Set1") + theme_bw()
```

5XFAD cerebral cortex samples cluster by genotype and age but not sex. The effects of genotype and age are much larger than any potential batch effects if the `NA` batch is a separate batch from the others. There might be a slight RIN bias between genotypes in a few groups.

## Quality plots of samples -- Hippocampus

```{r dds-h, message=FALSE, warning=FALSE}
meta_h <- subset(meta_vis, tissue == "hippocampus")
counts_h <- data$counts[, meta_h$merged_file_specimenID]

keep <- rowSums(counts_h > 10) >= 3

dds_h <- DESeqDataSetFromMatrix(counts_h[keep, ], meta_h, design = ~ 1)
dds_h <- DESeq(dds_h, quiet = TRUE)
plotDispEsts(dds_h)
```

```{r vst-h}
vsd_h <- vst(dds_h)
meanSdPlot(assay(vsd_h))
```

## PCA plots -- Hippocampus

```{r pca-h, message=FALSE}
meta_h$RIN <- as.numeric(meta_h$RIN)
create_pca_plot(meta_h, assay(vsd_h),
                title = "RIN",
                discrete_color = FALSE,
                brewer_palette = "OrRd",
                # aes specification
                color = RIN)

# Boxplots of RIN vs each group
meta_h$group <- paste(meta_h$sex, meta_h$ageDeath)
plot_grouped_boxplot(meta_h,
                     title = "RIN vs group",
                     # aes specification
                     x = group, y = RIN, color = genotype)

plotPCA(vsd_h, intgroup = c("genotype", "sex")) +
  ggtitle("Genotype vs sex") +
  scale_color_brewer(palette = "Set1") + theme_bw()

plotPCA(vsd_h, intgroup = c("genotype", "ageDeath")) +
  ggtitle("Genotype vs age") +
  scale_color_brewer(palette = "Set1") + theme_bw()

plotPCA(vsd_h, intgroup = c("sex", "ageDeath")) +
  ggtitle("Sex vs age") +
  scale_color_brewer(palette = "Set1") + theme_bw()

plotPCA(vsd_h, intgroup = c("rnaBatch", "ageDeath")) +
  ggtitle("Batch vs age") +
  scale_color_brewer(palette = "Set1") + theme_bw()

plotPCA(vsd_h, intgroup = c("rnaBatch", "genotype")) +
  ggtitle("Batch vs genotype") +
  scale_color_brewer(palette = "Set1") + theme_bw()
```

5XFAD hippocampus samples cluster by genotype and age but not sex, similar to the cerebral cortex. The effects of genotype and age are much larger than any batch effects. Given this, it may not be necessary to throw out samples from the NA batch for either tissue. There is no obvious bias toward RIN within each analysis group.

```{r upload, eval=FALSE, include=FALSE}
# Code to upload this file to Synapse and add its contents to its own wiki page,
# which makes it nicely viewable.
library(knit2synapse)
library(synapser)

synLogin(silent = TRUE)

# TODO add provenance
createAndKnitToFileEntity(file = "UCI_5XFAD_dataset.Rmd",
                          parentId = "syn64299959",
                          fileName = "UCI_5XFAD_dataset.Rmd",
                          wikiName = "UCI_5XFAD Dataset Exploration",
                          overwrite = TRUE,
                          forceVersion = FALSE)
```
