---
title: "UCI_Trem2-R47H_NSS Dataset Exploration"
author: Jaclyn Beck (Sage Bionetworks)
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Set up

```{r initialize, include=FALSE}
source("vis_utils.R")
data <- load_data("UCI_Trem2-R47H_NSS")
```

## Study design

```{r trem2-design}
meta_vis <- data$metadata %>%
  mutate(ageDeath = paste(ageDeath, "months"),
         sex = factor(sex))

# Study design
print(table(meta_vis$genotype, meta_vis$ageDeath, meta_vis$sex))
```

### Check for batches

7 batches present:

```{r trem2-batches}
data$metadata %>%
  group_by(rnaBatch, libraryBatch, sequencingBatch) %>%
  summarize(count = n(), .groups = "drop")
```

Genotypes are spread unevenly across batches:

```{r trem2-batches2}
meta_vis$batch <- paste(meta_vis$rnaBatch, meta_vis$sequencingBatch, 
                        meta_vis$libraryBatch, sep = "/")

table(meta_vis$batch, meta_vis$genotype)
```
Sexes are spread among all batches:

```{r trem2-batches3}
meta_vis$batch <- paste(meta_vis$rnaBatch, meta_vis$sequencingBatch, 
                        meta_vis$libraryBatch, sep = "/")

table(meta_vis$batch, meta_vis$sex)
```
Batches are specific to age group:

```{r trem2-batches4}
meta_vis$batch <- paste(meta_vis$rnaBatch, meta_vis$sequencingBatch, 
                        meta_vis$libraryBatch, sep = "/")

table(meta_vis$batch, meta_vis$ageDeath)
```
Looking at rnaBatch only:

```{r trem2-batches5}
table(meta_vis$rnaBatch, meta_vis$genotype, meta_vis$ageDeath)
```
5XFAD_carrier and 5XFAD_noncarrier mice are in batches 1 and 4, while the remaining two genotypes are in batches 2 and 3. Pairs of genotypes from the same age are in the same batch. 

Looking at libraryBatch only:

```{r trem2-batches6}
table(meta_vis$libraryBatch, meta_vis$genotype, meta_vis$ageDeath)
```
All genotypes for 12 month mice are in batches 2 and 3. For 4 month mice, 5XFAD_carrier and 5XFAD_noncarrier are in batch 1, while the remaining two genotypes are in batch 4.

Looking at sequencingBatch only:

```{r trem2-batches7}
table(meta_vis$sequencingBatch, meta_vis$genotype, meta_vis$ageDeath)
```
All 12 month mice are in batch 3. For 4 month mice, 5XFAD_carriers are in batch 1, 5XFAD_noncarriers are in 1 and 2, and the remaining two genotypes are in batch 4.

Final split of all batches by sex + age group:

```{r trem2-batches8}
table(meta_vis$batch, meta_vis$genotype, meta_vis$sex, meta_vis$ageDeath)
```

* For 12 month mice, the genotypes are paired as above in either batches 3-4 or 6-7, for all sexes.
* For 4 month mice, all Trem2-R47H-containing mice are in batch 5 for both sexes, while 5XFAD_carriers are in batch 1 and 5XFAD_noncarriers are in batches 1-2 for all sexes. This kind of split by genotype isn't ideal. 

### Human gene expression

```{r trem2-human-expression, fig.width=8}
plt <- plot_human_genes(data$counts, data$metadata, data$symbol_map)
print(plt)
```

### Quality plots of samples

```{r trem2-deseq, message=FALSE, warning=FALSE}
dds <- DESeqDataSetFromMatrix(data$counts, meta_vis, design = ~ genotype)
dds <- DESeq(dds, quiet = TRUE)
plotDispEsts(dds)
```

```{r trem2-vst}
vsd <- vst(dds, blind = FALSE)
meanSdPlot(assay(vsd))
```

### PCA plots

```{r trem2-pca, message=FALSE, warning=FALSE}
plotPCA(vsd, intgroup = "RIN") + theme_bw() + ggtitle("RIN")

plotPCA(vsd, intgroup = c("genotype", "sex")) + 
  ggtitle("Genotype vs sex") +
  scale_color_brewer(palette = "Set1") + theme_bw() 

plotPCA(vsd, intgroup = c("genotype", "ageDeath")) + 
  ggtitle("Genotype vs age") +
  scale_color_brewer(palette = "Set1") + theme_bw()

plotPCA(vsd, intgroup = c("sex", "ageDeath")) +
  ggtitle("Sex vs age") +
  scale_color_brewer(palette = "Set1") + theme_bw()
```

Trem2-R47H_NSS samples cluster by genotype (whether they contain 5XFAD or not) and age, but not sex.

```{r upload, eval=FALSE, include=FALSE}
# Code to upload this file to Synapse and add its contents to its own wiki page,
# which makes it nicely viewable.
library(knit2synapse)
library(synapser)

synLogin(silent = TRUE)

# TODO add provenance
createAndKnitToFileEntity(file = "UCI_Trem2-R47H_NSS_dataset.Rmd", parentId = "syn64299959", 
                          fileName = "UCI_Trem2-R47H_NSS_dataset.Rmd",
                          wikiName = "UCI_Trem2-R47H_NSS Dataset Exploration", 
                          overwrite = TRUE, forceVersion = FALSE)

```
